File: simpleGuessingGame.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <cstdlib>
#include <ctime>
#include <thread>
#include <chrono>
#include <limits>

#define RESET   "\033[0m"
#define RED     "\033[31m"
#define YELLOW  "\033[33m"

using namespace std;

int selectDifficulty() {
    cout << "\nSelect Difficulty Level:"
        "\n 1. Easy (1-10)"
        "\n 2. Medium (1-50)"
        "\n 3. Hard (1-100)"
        "\n";
    int choice;
    while (true) {
        cout << "Enter your choice: ";
        if (!(cin >> choice)) {
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            cout << RED << "\nInvalid input. Please enter a number.\n" << RESET;
            this_thread::sleep_for(chrono::seconds(1));
            continue;
        }
        if (choice >= 1 && choice <= 3)
            return choice;
        else {
            cout << RED << "\nInvalid Choice. Please try again.\n" << RESET;
            this_thread::sleep_for(chrono::seconds(1));
        }
    }
}

// Struct to store leaderboard entries
struct Entry {
    string name;
    int attempts;
};

// Read leaderboard file
vector<Entry> readLeaderboard(const string& filename) {
    vector<Entry> leaderboard;
    ifstream file(filename);
    if (!file.is_open()) return leaderboard; // file might not exist yet

    string line;
    getline(file, line); // skip "LEADERBOARD:"
    while (getline(file, line)) {
        if (line.empty()) continue;
        size_t commaPos = line.find(',');
        if (commaPos != string::npos) {
            string name = line.substr(0, commaPos);
            int tries = stoi(line.substr(commaPos + 1));
            leaderboard.push_back({name, tries});
        }
    }
    file.close();
    return leaderboard;
}

// Write leaderboard to file
void writeLeaderboard(const string& filename, vector<Entry>& leaderboard) {
    ofstream file(filename, ios::trunc);
    file << "LEADERBOARD:\n";
    for (const auto& entry : leaderboard) {
        file << entry.name << "," << entry.attempts << "\n";
    }
    file.close();
}

// Display leaderboard
void showLeaderboard(const vector<Entry>& leaderboard) {
    cout << "\n🏆 LEADERBOARD:\n";
    int rank = 1;
    for (const auto& entry : leaderboard) {
        cout << rank++ << ". " << entry.name << " - " << entry.attempts << " attempts\n";
    }
}

int main() {
    srand(static_cast<unsigned int>(time(nullptr)));
    cout << "Welcome to the Simple Guessing Game!\n";

    while (true) {
        int difficulty = selectDifficulty();
        int maxNumber;
        string filename;

        if (difficulty == 1) {
            maxNumber = 10;
            filename = "Results_Easy.txt";
        } else if (difficulty == 2) {
            maxNumber = 50;
            filename = "Results_Medium.txt";
        } else {
            maxNumber = 100;
            filename = "Results_Hard.txt";
        }

        int targetNumber = rand() % maxNumber + 1;
        int userGuess = 0;
        int attempts = 0;

        cout << "\nI have selected a number between 1 and " << maxNumber << ". Can you guess it?\n";

        while (true) {
            cout << "Enter your guess: ";
            if (!(cin >> userGuess)) {
                cin.clear();
                cin.ignore(numeric_limits<streamsize>::max(), '\n');
                cout << RED << "Please enter a valid number.\n" << RESET;
                continue;
            }

            attempts++;

            if (userGuess < 1 || userGuess > maxNumber) {
                cout << RED << "Invalid guess. Please enter between 1 and " << maxNumber << ".\n" << RESET;
                this_thread::sleep_for(chrono::seconds(1));
                continue;
            }

            if (userGuess < targetNumber) {
                cout << YELLOW << "Too low! Try again.\n" << RESET;
            } else if (userGuess > targetNumber) {
                cout << YELLOW << "Too high! Try again.\n" << RESET;
            } else {
                cout << "\n🎉 Congratulations! You guessed the number " << targetNumber 
                    << " in " << attempts << " attempts.\n";
                break;
            }
        }

        // Record to leaderboard
        cout << "Enter your name: ";
        string playerName;
        cin >> playerName;

        vector<Entry> leaderboard = readLeaderboard(filename);
        leaderboard.push_back({playerName, attempts});

        // Sort ascending by attempts
        sort(leaderboard.begin(), leaderboard.end(), [](const Entry& a, const Entry& b) {
            return a.attempts < b.attempts;
        });

        // Keep top 5
        if (leaderboard.size() > 5)
            leaderboard.resize(5);

        writeLeaderboard(filename, leaderboard);
        showLeaderboard(leaderboard);

        // Ask to play again
        char playAgain;
        cout << "\nPlay again? (y/n): ";
        cin >> playAgain;
        if (tolower(playAgain) != 'y') {
            cout << "\nThanks for playing! Goodbye!\n";
            break;
        }
    }

    return 0;
}

File: simpleGuessingGame.py
import random
import time
import os

class Colors:
    RESET = '\033[0m'
    RED = '\033[31m'
    YELLOW = '\033[33m'

def selectDifficulty():
    print("\nSelect Difficulty Level:")
    print("1. Easy (1-10)")
    print("2. Medium (1-50)")
    print("3. Hard (1-100)")
    
    while True:
        choice = input("\nEnter your choice (1/2/3): ").strip()
        if choice == '1':
            return 10
        elif choice == '2':
            return 50
        elif choice == '3':
            return 100
        else:
            print(Colors.RED + "\nInvalid choice. Please select 1, 2, or 3." + Colors.RESET)
            time.sleep(1)

def playGame():
    difficulty = selectDifficulty()
    randomNumber = random.randint(1, difficulty)
    attempts = 0

    print(f"\nI have selected a number between 1 and {difficulty}. Can you guess it?")

    while True:
        guess = input("\nEnter your guess: ").strip()
        attempts += 1
        
        # Validate numeric input
        try:
            guess = int(guess)
        except ValueError:
            print(Colors.YELLOW + "\nPlease enter a valid number." + Colors.RESET)
            time.sleep(1)
            continue
        
        # Check bounds
        if guess < 1 or guess > difficulty:
            print(Colors.RED + f"\nOut of bounds! Please guess between 1 and {difficulty}." + Colors.RESET)
            time.sleep(1)
            continue

        # Guess logic
        if guess < randomNumber:
            print("\nToo low! Try again.")
        elif guess > randomNumber:
            print("\nToo high! Try again.")
        else:
            print(f"\n🎉 Congratulations! You guessed the number {randomNumber} in {attempts} attempts.")
            player_name = input("Enter your name: ").strip() or "Anonymous"

            # Select leaderboard file based on difficulty
            if difficulty == 10:
                leaderboard_file = 'Results_Easy.txt'
            elif difficulty == 50:
                leaderboard_file = 'Results_Medium.txt'
            else:
                leaderboard_file = 'Results_Hard.txt'

            leaderboard = []

            # Load existing leaderboard
            if os.path.exists(leaderboard_file):
                with open(leaderboard_file, 'r') as file:
                    lines = file.readlines()
                    for line in lines[1:]:  # skip "LEADERBOARD:"
                        line = line.strip()
                        if line:
                            name, tries = line.split(',')
                            leaderboard.append((name, int(tries)))

            # Add and sort results
            leaderboard.append((player_name, attempts))
            leaderboard.sort(key=lambda x: x[1])
            leaderboard = leaderboard[:5]

            # Write updated leaderboard
            with open(leaderboard_file, 'w') as file:
                file.write("LEADERBOARD:\n")
                for name, tries in leaderboard:
                    file.write(f"{name},{tries}\n")

            # Display leaderboard
            print("\n🏆 LEADERBOARD:")
            for i, (name, tries) in enumerate(leaderboard, start=1):
                print(f"{i}. {name} - {tries} attempts")

            break  # Exit guessing loop

while True:
    playGame()
    option = input("\nDo you want to play again? (yes/no): ").strip().lower()
    if option == 'no':
        print("\nThank you for playing! Goodbye!")
        break
    elif option != 'yes':
        print(Colors.RED + "\nInvalid input. Please enter 'yes' or 'no'." + Colors.RESET)
